csrc/executor_kernel_arg.cpp

```diff
@@ -14,6 +14,292 @@
+    auto backward_exprs = StmtSort::getExprsBetween(
+        tv->fusion(),
+        {alloc.begin(), alloc.end()},
+        {rfactor.begin(), rfactor.end()});
+    std::reverse(backward_exprs.begin(), backward_exprs.end());
+    for (auto expr : backward_exprs) {
+      handle(expr);
+    }
+  }
+};
+
+} // namespace
+
+// Given an ATen tensor, whose sizes and strides are w.r.t to the rFactor domain
+// of its corresponding TensorView, compute the sizes and strides of the tensor
+// with respect to its allocation domain.
+// For example, if the rFactor domain is [I1, I2], and the allocation domain is
+// [I2*I1], and the tensor's size is [5, 3] and stride is [2, 10], then the
+// resulting size will be [15] and stride will be [2]
+// Another example, if the rFactor domain is [I1*I2] and the allocation domain
+// is [I1, I2], and the tensor's size is [15] and stride is [7], and the extent
+// of I2 is 5, then the resulting size will be [3, 5] and stride will be [35, 7]
+std::vector<std::pair<int64_t, int64_t>>
+inferAndValidateAllocationSizesAndStrides(
+    const at::Tensor& tensor,
+    TensorView* tv,
+    ExpressionEvaluator& ee) {
+  if (tv == nullptr || !tv->hasAllocation()) {
+    // When tv is nullptr, or tv does not have allocation, the given sizes and
+    // strides should already be in the target format. So nothing to do here.
+    std::vector<std::pair<int64_t, int64_t>> result;
+    for (auto i : c10::irange(tensor.dim())) {
+      result.emplace_back(tensor.size(i), tensor.stride(i));
+    }
+    return result;
+  }
+  const auto& alloc =
+      TensorDomain::noReductions(tv->getMaybeAllocationDomain());
+  const auto& rfactor = TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+
+  // active IDs and their shape and stride
+  std::unordered_map<IterDomain*, std::pair<int64_t, int64_t>> active_ids;
+  TORCH_INTERNAL_ASSERT((int64_t)rfactor.size() == tensor.dim());
+  for (int64_t i : c10::irange((int64_t)rfactor.size())) {
+    auto rf_id = rfactor.at(i);
+    active_ids[rf_id] = {tensor.size(i), tensor.stride(i)};
+  }
+
+  ForwardTraverseFromRFactorToAlloc(ee, active_ids).run(tv, rfactor, alloc);
+  BackwardTraverseFromRFactorToAlloc(ee, active_ids).run(tv, rfactor, alloc);
+
+  // Now active_ids should contain the final sizes and strides, unordered. We
+  // need to put them to the correct order.
+  std::vector<std::pair<int64_t, int64_t>> sizes_strides;
+  sizes_strides.reserve(alloc.size());
+  for (auto i : c10::irange(alloc.size())) {
+    auto id = alloc.at(i);
+    sizes_strides.emplace_back(active_ids.at(id));
+  }
+  // Validate final sizes and strides with contiguity
+  int64_t contiguous_stride = 1;
+  std::vector<std::optional<bool>> contiguity = tv->getContiguity();
+  for (int64_t i = (int64_t)sizes_strides.size() - 1; i >= 0; i--) {
+    if (alloc.at(i)->isBroadcast()) {
+      continue;
+    }
+    while (!contiguity.back().has_value()) {
+      contiguity.pop_back();
+      continue;
```

Comment:
This `continue` seems redundant.

Reply:
Actually, not really sure what this while loop is trying to do.

Reply:
`continue` removed. The size of `contiguity` does not necessarily equal to the size of `sizes_strides`, because reductions do have contiguity but are not in `sizes_strides`. So this loop just pop all null contiguity on the back, these null contiguity might be from a reduction or from a broadcast of the last iteration of the outer loop. Whatever this null contiguity comes from, it does not belong to this iteration, so it should be just skipped.

Reply:
I added a `TORCH_INTERNAL_ASSERT(contiguity.empty(),` to then end of the outer loop to check if the "actual size" of `contiguity` matches with `sizes_strides`.